Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in Source/WebCore/platform/audio

cdumez opened the pull request:
#### 3d996666c89c13d17718f5ee0ff71f9d97f36c9b
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in Source/WebCore/platform/audio
<a href="https://bugs.webkit.org/show_bug.cgi?id=285203">https://bugs.webkit.org/show_bug.cgi?id=285203</a>

Reviewed by Geoffrey Garen and Darin Adler.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::reinterpretCastSpanStartTo):
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestinationCocoa::render):
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::validateAudioBufferList):
(WebCore::passthroughInputDataCallback):
(WebCore::AudioFileReader::decodeWebMData const):
(WebCore::AudioFileReader::createBus):
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
(WebCore::AudioSampleBufferList::applyGain):
(WebCore::mixBuffers):
(WebCore::AudioSampleBufferList::copyTo):
(WebCore::AudioSampleBufferList::zeroABL):
(WebCore::audioConverterFromABLCallback):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::CARingBuffer::CARingBuffer):
(WebCore::CARingBuffer::initialize):
(WebCore::ZeroRange):
(WebCore::StoreABL):
(WebCore::FetchABL):
(WebCore::ZeroABL):
(WebCore::CARingBuffer::store):
(WebCore::CARingBuffer::fetchInternal):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.h:
(WebCore::CARingBuffer::span):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::supportsSpatialAudioPlaybackForConfiguration):
* Source/WebCore/platform/audio/cocoa/SpanCoreAudio.h:
(WebCore::dataSpan):
(WebCore::dataMutableSpan):
(WebCore::span):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
(WebCore::WebAudioBufferList::WebAudioBufferList):
(WebCore::WebAudioBufferList::setSampleCount):
(WebCore::WebAudioBufferList::setSampleCountWithBlockBuffer):
(WebCore::WebAudioBufferList::initializeList):
(WebCore:: const):
(WebCore::WebAudioBufferList::buffer const):
* Source/WebCore/platform/audio/mac/AudioHardwareListenerMac.cpp:
(WebCore::AudioHardwareListenerMac::AudioHardwareListenerMac):
(WebCore::AudioHardwareListenerMac::propertyChanged):
* Source/WebCore/platform/audio/mac/AudioHardwareListenerMac.h:
* Source/WebCore/platform/audio/mac/AudioSessionMac.mm:
(WebCore::AudioSessionMac::maximumNumberOfOutputChannels const):
* Source/WebCore/platform/audio/mac/FFTFrameMac.cpp:
(WebCore::FFTFrame::doFFT):
(WebCore::FFTFrame::doInverseFFT):

Canonical link: <a href="https://commits.webkit.org/288322@main">https://commits.webkit.org/288322@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e7316962f1398aa5dfa2de620f60fa25cd9b6b1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82478 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2139 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36569 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87608 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33539 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84580 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2210 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10033 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64282 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22038 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85544 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1578 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75036 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44559 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1476 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29223 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32577 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75466 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72768 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29861 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88967 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81532 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9781 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7043 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72685 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10008 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70850 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71902 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16042 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15049 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12804 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9734 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15255 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103942 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9608 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25217 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13074 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11378 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
```diff
@@ -113,8 +111,8 @@ OSStatus AudioDestinationCocoa::render(double sampleTime, uint64_t hostTime, UIn
 
-    auto* buffers = ioData->mBuffers;
     auto numberOfBuffers = std::min<UInt32>(ioData->mNumberBuffers, m_outputBus->numberOfChannels());
+    auto buffers = span(*ioData).first(numberOfBuffers);
```
Why the call to `first`?

cdumez replied:
It's not strictly needed since the loop below will loop `numberOfBuffers` times anyway. I'll remove.

geoffreygaren commented on Source/WebCore/platform/audio/cocoa/SpanCoreAudio.h:
```diff
@@ -51,4 +51,26 @@ inline std::span<const float> dataFloatSpan(AudioBuffer& buffer)
+}
+
+template<typename T>
+inline std::span<T> dataMutableSpan(AudioBuffer& buffer)
```
I think this reads better as `mutableDataSpan`.

geoffreygaren commented on Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
```diff
@@ -360,11 +361,14 @@ std::optional<size_t> AudioFileReader::decodeWebMData(AudioBufferList& bufferLis
+            for (UInt32 i = 0; i < inFormat.mChannelsPerFrame; ++i) {
+                decodedBuffers[i].mNumberChannels = 1;
+                decodedBuffers[i].mDataByteSize = numFrames * sizeof(float);
+                decodedBuffers[i].mData = dataMutableFloatSpan(bufferListBuffers[i]).subspan(decodedFrames).data();
```
Maybe this was discussed in past review, but I think `dataMutableFloatSpan` would read better as `mutableDataFloatSpan`. We could also drop the 'data', since no other span is possible: `mutableFloatSpan`.

geoffreygaren commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
```diff
@@ -66,31 +65,30 @@ void AudioSampleBufferList::setSampleCount(size_t count)
-            int frameCount = bufferList.mBuffers[i].mDataByteSize / sizeof(int32_t);
-            for (int i = 0; i < frameCount; i++)
-                buffer[i] *= gain;
+            auto data = dataMutableSpan<int32_t>(buffer);
```
`mutableSpan`

geoffreygaren commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
```diff
@@ -112,40 +110,42 @@ void AudioSampleBufferList::applyGain(float gain)
-            auto* source = static_cast<int16_t*>(sourceBuffer.mBuffers[i].mData);
-            for (size_t i = 0; i < frameCount; i++)
-                destination[i] += source[i];
+            auto sourceData = dataSpan<int16_t>(source);
```
`span`

geoffreygaren commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
```diff
@@ -112,40 +110,42 @@ void AudioSampleBufferList::applyGain(float gain)
         case AudioStreamDescription::Int32: {
-            ASSERT(frameCount <= destinationBuffer.buffer(i)->mDataByteSize / 4);
-            ASSERT(frameCount <= sourceBuffer.mBuffers[i].mDataByteSize / 4);
+            ASSERT(frameCount <= destination.mDataByteSize / 4);
```
I think `4` would read more clearly as `sizeof(int32_t)`. (I think that's what we're trying to match here?)

geoffreygaren commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
```diff
@@ -112,40 +110,42 @@ void AudioSampleBufferList::applyGain(float gain)
-            auto* destination = static_cast<int32_t*>(destinationBuffer.buffer(i)->mData);
-            vDSP_vaddi(destination, 1, reinterpret_cast<int32_t*>(sourceBuffer.mBuffers[i].mData), 1, destination, 1, frameCount);
+            auto* destinationPtr = static_cast<int32_t*>(destination.mData);
+            vDSP_vaddi(destinationPtr, 1, reinterpret_cast<int32_t*>(source.mData), 1, destinationPtr, 1, frameCount);
```
Technically this is a violation of bounds safety -- though it is our callee that does the unbounded dereference, rather than our own code. We have ASSERTs, but those don't do anything in release builds.

The most principled thing to do would be to create wrappers like `vDSP_vaddi_span` that took spans instead of pointers, and RELEASE_ASSERTed that all three spans were big enough for frameCount before calling through to vDSP_vaddi. That's the moral equivalent of what we do with memcpySpan.

cdumez replied:
OK, I'll add a wrapper like so to VectorMath.cpp:
```
void add(std::span<const int> inputVector1, std::span<const int> inputVector2, std::span<int> outputVector)
{
    RELEASE_ASSERT(inputVector1.size() == inputVector2.size());
    RELEASE_ASSERT(outputVector.size() >= inputVector1.size());
    vDSP_vaddi(inputVector1.data(), 1, inputVector2.data(), 1, outputVector.data(), 1, inputVector1.size());
}
```

geoffreygaren commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
```diff
@@ -112,40 +110,42 @@ void AudioSampleBufferList::applyGain(float gain)
-            auto* destination = static_cast<float*>(destinationBuffer.buffer(i)->mData);
-            vDSP_vadd(destination, 1, reinterpret_cast<float*>(sourceBuffer.mBuffers[i].mData), 1, destination, 1, frameCount);
+            auto* destinationPtr = static_cast<float*>(destination.mData);
+            vDSP_vadd(destinationPtr, 1, reinterpret_cast<float*>(source.mData), 1, destinationPtr, 1, frameCount);
```
Same comment as `vDSP_vaddi`.

geoffreygaren commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
```diff
@@ -112,40 +110,42 @@ void AudioSampleBufferList::applyGain(float gain)
-            auto* destination = static_cast<double*>(destinationBuffer.buffer(i)->mData);
-            vDSP_vaddD(destination, 1, reinterpret_cast<double*>(sourceBuffer.mBuffers[i].mData), 1, destination, 1, frameCount);
+            auto* destinationPtr = static_cast<double*>(destination.mData);
+            vDSP_vaddD(destinationPtr, 1, reinterpret_cast<double*>(source.mData), 1, destinationPtr, 1, frameCount);
```
Same comment as `vDSP_vaddi`.

geoffreygaren commented on Source/WebCore/platform/audio/mac/FFTFrameMac.cpp:
```diff
@@ -107,9 +107,7 @@ FFTFrame::~FFTFrame() = default;
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-    vDSP_ctoz(reinterpret_cast<const DSPComplex*>(data.data()), 2, &m_frame, 1, halfSize);
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+    vDSP_ctoz(&reinterpretCastSpanStartTo<const DSPComplex>(data), 2, &m_frame, 1, halfSize);
```
Same comment as `vDSP_vaddi`.

geoffreygaren commented on Source/WebCore/platform/audio/mac/FFTFrameMac.cpp:
```diff
@@ -127,9 +125,7 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-    vDSP_ztoc(&m_frame, 1, reinterpret_cast<DSPComplex*>(data.data()), 2, m_FFTSize / 2);
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+    vDSP_ztoc(&m_frame, 1, &reinterpretCastSpanStartTo<DSPComplex>(data), 2, m_FFTSize / 2);
```
Same comment as `vDSP_vaddi`.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in Source/WebCore/platform/audio
https://bugs.webkit.org/show_bug.cgi?id=285203

Reviewed by Geoffrey Garen and Darin Adler.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::reinterpretCastSpanStartTo):
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestinationCocoa::render):
* Source/WebCore/platform/audio/cocoa/AudioFileReaderCocoa.cpp:
(WebCore::validateAudioBufferList):
(WebCore::passthroughInputDataCallback):
(WebCore::AudioFileReader::decodeWebMData const):
(WebCore::AudioFileReader::createBus):
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp:
(WebCore::AudioSampleBufferList::applyGain):
(WebCore::mixBuffers):
(WebCore::AudioSampleBufferList::copyTo):
(WebCore::AudioSampleBufferList::zeroABL):
(WebCore::audioConverterFromABLCallback):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::CARingBuffer::CARingBuffer):
(WebCore::CARingBuffer::initialize):
(WebCore::ZeroRange):
(WebCore::StoreABL):
(WebCore::FetchABL):
(WebCore::ZeroABL):
(WebCore::CARingBuffer::store):
(WebCore::CARingBuffer::fetchInternal):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.h:
(WebCore::CARingBuffer::span):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::supportsSpatialAudioPlaybackForConfiguration):
* Source/WebCore/platform/audio/cocoa/SpanCoreAudio.h:
(WebCore::dataSpan):
(WebCore::dataMutableSpan):
(WebCore::span):
* Source/WebCore/platform/audio/cocoa/WebAudioBufferList.cpp:
(WebCore::WebAudioBufferList::WebAudioBufferList):
(WebCore::WebAudioBufferList::setSampleCount):
(WebCore::WebAudioBufferList::setSampleCountWithBlockBuffer):
(WebCore::WebAudioBufferList::initializeList):
(WebCore:: const):
(WebCore::WebAudioBufferList::buffer const):
* Source/WebCore/platform/audio/mac/AudioHardwareListenerMac.cpp:
(WebCore::AudioHardwareListenerMac::AudioHardwareListenerMac):
(WebCore::AudioHardwareListenerMac::propertyChanged):
* Source/WebCore/platform/audio/mac/AudioHardwareListenerMac.h:
* Source/WebCore/platform/audio/mac/AudioSessionMac.mm:
(WebCore::AudioSessionMac::maximumNumberOfOutputChannels const):
* Source/WebCore/platform/audio/mac/FFTFrameMac.cpp:
(WebCore::FFTFrame::doFFT):
(WebCore::FFTFrame::doInverseFFT):

Canonical link: https://commits.webkit.org/288322@main
